File: slices/gatekeeper/gatekeeper-core/src/page-paths.ts
// oxlint-disable import/group-exports -- Exports are already namespaced
/**
 * Single source of truth for the owner-UI routes of the gatekeeper's
 * browser-facing pages. The SPA router in `gatekeeper-react` declares matching
 * `<Route path>` values; a drift test asserts they agree.
 *
 * @remarks
 * Each is an owner-UI-relative path; path params are percent-encoded — callers
 * MUST NOT re-encode. The absolute URLs a server hands a browser are built on
 * the Rust side (`gatekeeper-rust`'s `domain/page_paths.rs`): the hosted owner
 * UI's base plus this path, with `?server=<served origin>` naming the server
 * the page should talk back to.
 */
export namespace GatekeeperPaths {
  export const oauthConsentPath = (id: string): string =>
    `/gatekeeper/oauth-consent/${encodeURIComponent(id)}`

  export const deviceEntryPath = (): string => `/gatekeeper/devices`

  /**
   * The device-entry page a server's `verification_uri` (or
   * `verification_uri_complete`) names, moved onto the owner UI copy served at
   * `servedRoot` — keeping the query the server built: `?server=` naming the
   * origin it was reached at, and `user_code` on the complete form. The server
   * builds its URIs on the owner UI it is configured with; a copy of the owner
   * UI showing a pairing it started itself (a PR preview, a dev server) links to
   * its own device-entry page instead, so the pairing is approved on the same
   * copy. The server is not asked to trust any address for this.
   *
   * `servedRoot` is slash-terminated (the copy's origin plus its basepath).
   * Throws a `TypeError` when `verificationUri` is not an absolute URL.
   *
   * @example GatekeeperPaths.deviceEntryUrlOn('https://x.test/pr-7/app/', 'https://wildflowerhealth.io/app/gatekeeper/devices?server=s&user_code=AB') // 'https://x.test/pr-7/app/gatekeeper/devices?server=s&user_code=AB'
   */
  export const deviceEntryUrlOn = (servedRoot: string, verificationUri: string): string => {
    const page = new URL(deviceEntryPath().replace(/^\//, ''), servedRoot)
    page.search = new URL(verificationUri).search
    return page.href
  }

  export const deviceConsentPath = (userCode: string): string =>
    `/gatekeeper/devices/${encodeURIComponent(userCode)}`
}

File: slices/gatekeeper/gatekeeper-rust/src/page_paths.rs
use url::form_urlencoded;

pub fn oauth_polling_path(id: &str) -> String {
    format!("/gatekeeper/oauth-polling/{}", percent_encode(id))
}

pub fn oauth_consent_path(id: &str) -> String {
    format!("/gatekeeper/oauth-consent/{}", percent_encode(id))
}

pub fn device_entry_path() -> &'static str {
    "/gatekeeper/devices"
}

pub fn device_consent_path(user_code: &str) -> String {
    format!("/gatekeeper/devices/{}", percent_encode(user_code))
}

pub fn oauth_polling_url(origin: &str, id: &str) -> String {
    format!("{origin}{}", oauth_polling_path(id))
}

pub fn oauth_consent_url(origin: &str, id: &str) -> String {
    format!("{origin}{}", oauth_consent_path(id))
}

pub fn device_entry_url(origin: &str) -> String {
    format!("{origin}{}", device_entry_path())
}

pub fn device_entry_url_with_code(origin: &str, user_code: &str) -> String {
    let query: String = form_urlencoded::Serializer::new(String::new())
        .append_pair("user_code", user_code)
        .finish();
    format!("{origin}{}?{}", device_entry_path(), query)
}

pub fn device_consent_url(origin: &str, user_code: &str) -> String {
    format!("{origin}{}", device_consent_path(user_code))
}

fn percent_encode(s: &str) -> String {
    form_urlencoded::byte_serialize(s.as_bytes())
        .collect::<String>()
        .replace('+', "%20")
}
